File: 290-word-pattern/word-pattern.js
/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
var wordPattern = function(pattern, s) {
  const words = s.split(" ");

  if (pattern.length !== words.length) {
    return false;
  }

  const patternToWordMap = new Map();
  const seenWords = new Set();

  for (let i = 0; i < pattern.length; i++) {
    const currentPatternChar = pattern[i];
    const currentWord = words[i];

    if (patternToWordMap.has(currentPatternChar)) {
      if (patternToWordMap.get(currentPatternChar) !== currentWord) {
        return false;
      }
    } else {
      if (seenWords.has(currentWord)) {
        return false;
      }

      patternToWordMap.set(currentPatternChar, currentWord);
      seenWords.add(currentWord);
    }
  }

  return true;
};

File: 290-word-pattern/word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split(" ")

        if len(pattern) != len(words):
            return False

        pattern_to_word_map = {}
        seen_words = set()

        for char, word in zip(pattern, words):
            if char not in pattern_to_word_map:
                if word in seen_words:
                    return False

                pattern_to_word_map[char] = word
                seen_words.add(word)
            elif pattern_to_word_map[char] != word:
                return False

        return True
